File: app/src/main/java/org/zwsmith/myapplication/core/repositories/UserRepository.java
package org.zwsmith.myapplication.core.repositories;

import android.support.annotation.NonNull;
import android.util.Log;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import org.zwsmith.myapplication.core.models.User;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class UserRepository {
    private static final String TAG = UserRepository.class.getSimpleName();
    private final DatabaseReference databaseReference = FirebaseDatabase.getInstance().getReference();

    private User currentUser = null;

    @Inject
    public UserRepository() {
    }

    public void createUser(String userId, String displayName, String email) {
        final User newUser = new User(displayName, email);
        final DatabaseReference userReference = databaseReference.child("users").child(userId);

        userReference.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {
                    User existingUser = dataSnapshot.getValue(User.class);
                    Log.i(TAG, "User already exists: " + existingUser.username);
                } else {
                    userReference
                            .setValue(newUser)
                            .addOnSuccessListener(new OnSuccessListener<Void>() {
                                @Override
                                public void onSuccess(Void aVoid) {
                                    Log.i(TAG, "Added newUser: " + newUser.email);
                                }
                            });
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                Log.d(TAG, "Database error: " + databaseError.getMessage());
            }
        });
    }
}

File: app/src/main/java/org/zwsmith/myapplication/core/repositories/UserRepository.kt
package org.zwsmith.myapplication.core.repositories

import android.util.Log
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.FirebaseFirestore

import org.zwsmith.myapplication.core.models.User

import javax.inject.Inject
import javax.inject.Singleton
import kotlin.Exception

@Singleton
class UserRepository @Inject constructor() {
    private val databaseReference = FirebaseFirestore.getInstance()

    private val currentUser: User? = null

    fun createUser(firebaseUser: FirebaseUser) {
        val user = hashMapOf<String, Any?>()
                .apply {
                    put(DISPLAY_NAME_KEY, firebaseUser.displayName)
                    put(EMAIL_KEY, firebaseUser.email)
                    put(WORKOUT_COUNT, 0)
                }

        databaseReference.collection(USERS_COLLECTION_KEY)
                .document(firebaseUser.uid)
                .set(user)
                .addOnSuccessListener {
                    Log.i(TAG, "New user added")
                }
                .addOnFailureListener { e: Exception ->
                    Log.e(TAG, "Error creating user: ${e.message}", e)
                }
    }

    companion object {
        private val TAG = UserRepository::class.java.simpleName
        private const val DISPLAY_NAME_KEY = "display_name"
        private const val EMAIL_KEY = "email"
        private const val WORKOUT_COUNT = "workout_count"
        private const val USERS_COLLECTION_KEY = "users"
    }
}
